staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -158,25 +171,57 @@ func (q *delayingType) ShutDown() {
 	})
 }
 
+// Add sends an item for immediate queuing, preempting a waiting item as it is queued.
+func (q *delayingType) Add(item interface{}) {
+	q.addToDelayedQueue(item, 0, false)
+}
+
 // AddAfter adds the given item to the work queue after the given delay
 func (q *delayingType) AddAfter(item interface{}, duration time.Duration) {
+	q.addToDelayedQueue(item, duration, false)
+}
+
+// AddIfNotWaiting sends an item for immediate queuing, being dropped if the same item is already waiting.
+func (q *delayingType) AddIfNotWaiting(item interface{}) {
+	q.addToDelayedQueue(item, 0, true)
+}
+
+// AddIfNotWaitingAfter adds the given item to the work queue after the given delay, being dropped if the same
+// item is already waiting.
+func (q *delayingType) AddIfNotWaitingAfter(item interface{}, duration time.Duration) {
+	q.addToDelayedQueue(item, duration, true)
+}
+
+// addToDelayedQueue contains the implementation of the different add methods for the queue.
+func (q *delayingType) addToDelayedQueue(item interface{}, duration time.Duration, dw bool) {
 	// don't add if we're already shutting down
 	if q.ShuttingDown() {
 		return
 	}
 
-	q.metrics.retry()
+	// Add the item into the waiting queue with the appropriate ready time
+	waitEntry := &waitFor{data: item, readyAt: q.clock.Now().Add(duration), dropIfWaiting: dw}
 
-	// immediately add things with no delay
-	if duration <= 0 {
-		q.Add(item)
-		return
+	// For immediate Add's add a call-back channel to the entry so that we can block on it being processed.
```

Comment:
huh, this is clever but really unfortunate, it seems like a lot of extra waiting without much benefit (also I try to avoid test code in production code). What tests fail if this isn't here?

Reply:
Hmm, yes it appears that quite a lot or most controllers rely on Add and AddAfter with 0 delay to be synchronous for testing because they can’t poll the waiting channel to check it has cleared like the tests within the package.

I think I have another idea which could allow tests to call a synchronous Add or AddAfter but prevent the regular (default) Add from having to wait.  It’ll mean slightly updating all of the tests.

Reply:
Sorry we keep hitting road blocks on this, if it's really hard to change the tests maybe we can get a second opinion from @liggitt

Reply:
I'd be *really* hesitant to change behavior for existing users, whether they are tests or not. I wouldn't assume the immediate-add behavior is only test-facing

Reply:
Ok, I think that does mean that we need the immediate Add() and the AddAfter() to be synchronous even though the AddAfter behaviour with any other delay is not.

Reply:
I'm not convinced, Jordan - the point of this PR is actually to change behavior, specifically to not do things twice accidentally.

Reply:
I'd be OK with merging with this extra wait, then taking it out and fixing tests in a subsequent PR? That way we can be more certain if it is a problem or not.

Reply:
> I'm not convinced, Jordan - the point of this PR is actually to change behavior, specifically to not do things twice accidentally.

oh, based on the title, I thought this was adding new behavior/features... I thought the modified behavior for current users was a side effect of the refactoring to accomplish that

making AddAfter a no-op if the object is already queued seems like a plausible change

making Add pre-empty queued delayed adds seems like a plausible change

making Add and AddAfter(0) change from synchronous to asynchronous seems more questionable

Reply:
There's two queues; Add / AddAfter(0) were never synchronous wrt actually starting execution immediately, they just got into the pending execution queue in one step. After this change, they still skip the delaying queue but are serialized with all other queue operations.

The serialization bit is kind of necessary because the double-add bugs involved things getting added to the pending execution queue without checking the delaying queue.

Reply:
I originally started by looking at locks but such as solution might impact performance because we would be losing the 1000 wide channel feeding the waitingLoop.  The solution as it now stands is a compromise, there is a slight additional overhead on Add or AddAfter (with 0 delay) calls but only as long as it takes to feed the entries into the waitingLoop, but if a client wants maximum performance and is not bothered about having a synchronous add they could call the new AddWithOptions() method and specific the AsyncAdd option to get maximum performance - no locking, as fast as it takes to get the item into the waitingLoop feeder channel and return.

Reply:
I've updated the channel to indicate that we do in fact want to change the default behaviour.